source/opt/decoration_manager.cpp

```diff
@@ -39,25 +40,81 @@ std::vector<const ir::Instruction*> DecorationManager::GetDecorationsFor(
       ->InternalGetDecorationsFor<const ir::Instruction*>(id, include_linkage);
 }
 
-// TODO(pierremoreau): The code will return true for { deco1, deco1 }, { deco1,
-//                     deco2 } when it should return false.
 bool DecorationManager::HaveTheSameDecorations(uint32_t id1,
                                                uint32_t id2) const {
-  const auto decorationsFor1 = GetDecorationsFor(id1, false);
-  const auto decorationsFor2 = GetDecorationsFor(id2, false);
-  if (decorationsFor1.size() != decorationsFor2.size()) return false;
-
-  for (const ir::Instruction* inst1 : decorationsFor1) {
-    bool didFindAMatch = false;
-    for (const ir::Instruction* inst2 : decorationsFor2) {
-      if (AreDecorationsTheSame(inst1, inst2, true)) {
-        didFindAMatch = true;
-        break;
-      }
-    }
-    if (!didFindAMatch) return false;
-  }
-  return true;
+  using InstructionList = std::vector<const ir::Instruction*>;
+  using UniqueSortedList = std::vector<std::u32string>;
+
+  const InstructionList decorationsFor1 = GetDecorationsFor(id1, false);
+  const InstructionList decorationsFor2 = GetDecorationsFor(id2, false);
+
+  // This function will sort and remove all duplicates for each of the three
+  // given lists.
+  const auto generateUniqueSortedLists =
+      [](const InstructionList& decorationList, UniqueSortedList* decorateList,
+         UniqueSortedList* decorateIdList,
+         UniqueSortedList* memberDecorateList) {
+        std::unordered_set<std::u32string> decorateListTmp;
+        std::unordered_set<std::u32string> decorateIdListTmp;
+        std::unordered_set<std::u32string> memberDecorateListTmp;
+
+        for (const ir::Instruction* inst : decorationList) {
+          std::vector<uint32_t> binaryInst;
+          inst->ToBinaryWithoutAttachedDebugInsts(&binaryInst);
+
+          // Remove the target as we do not want to have them being compared.
+          binaryInst.erase(binaryInst.begin() + 1);
+
+          switch (inst->opcode()) {
+            case SpvOpDecorate:
+              decorateListTmp.emplace(
+                  reinterpret_cast<char32_t*>(binaryInst.data()),
+                  binaryInst.size());
+              break;
+            case SpvOpMemberDecorate:
+              memberDecorateListTmp.emplace(
+                  reinterpret_cast<char32_t*>(binaryInst.data()),
+                  binaryInst.size());
+              break;
+            case SpvOpDecorateId:
+              decorateIdListTmp.emplace(
+                  reinterpret_cast<char32_t*>(binaryInst.data()),
+                  binaryInst.size());
+              break;
+            default:
+              break;
+          }
+        }
+        std::copy(decorateListTmp.begin(), decorateListTmp.end(),
```

Comment:
I do need to sort them, otherwise it says that `{ Decorate %1 Constant, Decorate %1 Restrict }` and `{ Decorate %1 Restrict, Decorate %1 Constant }` are different, which is not the expected behaviour.

Reply:
If sets have the same content, then operator== would return true regardless of insertion order.

Reply:
Thanks for reminding me about std::set… I should have gone for that rather than emulating them with that unordered_map + vector.

Reply:
> If sets have the same content, then operator== would return true regardless of insertion order.

I meant for unordered sets, not ordered ones.

Reply:
operator== would have worked the same for unordered_set as for set, I'm not sure why you were having problems.

If you are using set, you can also use std::vector<uint32_t> instead of u32string.